hpx/parallel/util/numa_binding_allocator.hpp

```diff
@@ -0,0 +1,569 @@
+///////////////////////////////////////////////////////////////////////////////
+//  Copyright (c) 2017-2019 John Biddiscombe
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+///////////////////////////////////////////////////////////////////////////////
+
+#ifndef HPX_COMPUTE_HOST_NUMA_BINDING_ALLOCATOR_HPP
+#define HPX_COMPUTE_HOST_NUMA_BINDING_ALLOCATOR_HPP
+
+#include <hpx/config.hpp>
+
+#include <hpx/parallel/execution_policy.hpp>
+#include <hpx/runtime/threads/topology.hpp>
+
+#include <hpx/runtime/threads/executors/pool_executor.hpp>
+#include <hpx/runtime/threads/executors/guided_pool_executor.hpp>
+
+#include <cstddef>
+#include <type_traits>
+#include <utility>
+#include <iostream>
+#include <memory>
+
+#if defined(__linux) || defined(linux) || defined(__linux__)
+# include <linux/unistd.h>
+# include <sys/mman.h>
+# define NUMA_ALLOCATOR_LINUX
+#endif
+
+// Can be used to enable debugging of the allocator page mapping
+//#define NUMA_BINDING_ALLOCATOR_INIT_MEMORY
+// #define NUMA_BINDING_ALLOCATOR_DEBUG_PAGE_BINDING
+
+#if defined(NUMA_BINDING_ALLOCATOR_DEBUG_PAGE_BINDING) && !defined(HPX_MSVC)
+# include <plugins/parcelport/parcelport_logging.hpp>
+# define LOG_NUMA_MSG(x)                                   \
+    std::cout << "<NUMA> " << THREAD_ID << ' ' << x << "\n"
+#else
+# define LOG_NUMA_MSG(x)
+#endif
+
+namespace hpx { namespace threads { namespace executors
+{
+    struct numa_binding_allocator_tag {};
+
+    template <>
+    struct HPX_EXPORT pool_numa_hint<numa_binding_allocator_tag> {
+      // The call operator () must return an int type
+      // The arguments must be const ref versions of the equivalent task arguments
+      int operator()(const std::size_t& domain) const {
+          LOG_NUMA_MSG("allocator pool_numa_hint returns domain " << domain);
+          return domain;
```

Comment:
This will cause `size_t -> int` conversion warnings on some platforms (implicit narrowing conversion).

Reply:
I'll fix it - thanks.